Update "What's New" and "Glossary" (#404)
amzn/ion-docs: _books/ion-1-1/src/whats_new.md

Data density is another motivation. Certain encodings (e.g., timestamps, 
integers) have been made more compact and efficient. More significantly, macros now enable applications to have very
flexible interning of their data's structure. In aggregate, data transcoded from Ion 1.0 to Ion 1.1 should be more
compact and more efficient to both read and write.

## Backwards compatibility

Ion 1.0 and Ion 1.1 share the same data model.
Any data that can be represented in Ion 1.0 can also be represented with full fidelity in Ion 1.1 and vice versa.
This means that it is always possible to convert data from one version to the other without risk of data loss.

Ion 1.1 readers should be able to understand both Ion 1.0 and Ion 1.1 data.

The text encoding grammar of Ion 1.1 is a superset of Ion 1.0's text encoding grammar.
Any Ion 1.0 text data can also be parsed by an Ion 1.1 text parser.

Ion 1.1's binary encoding is substantially different from Ion 1.0's binary encoding.
Many changes have been made to make values more compact, faster to read and/or faster to write.

> Ion 1.0's [type descriptors](https://amazon-ion.github.io/ion-docs/docs/binary.html#typed-value-formats)

Reviewer: I can't place this comment in the right spot due do limitations in the PR UI. This pertains to the line below:

> Two streams using different Ion versions can be safely concatenated together provided that they are both text or both binary.
A concatenated stream containing both Ion 1.0 and Ion 1.1 can only be fully read by a reader that supports Ion 1.1.

I think this needs one more caveat: the concatenated segments *must* be preceded by an IVM matching the version of the encoding that follows. It's common for Ion 1.0 text to elide the Ion 1.0 IVM, but this won't necessarily work when concatenated after an Ion 1.1 stream due to Ion 1.1's no-op handling of Ion 1.0-style symbol tables.

## Macros, templates, and encoding expressions

Ion 1.1 introduces a new primitive called an *encoding expression* (*E-expression*). These expressions are (in text
syntax) similar to S-expressions, but they are not part of the data model.
E-Expressions represent encoding details, and can be used to define macros, invoke macros, and modify the encoding context.

> Template e-expressions and are _evaluated_ into one Ion value, which enables compact representation of Ion data. 

Reviewer: ```suggestion
Template e-expressions are _evaluated_ into one Ion value, which enables compact representation of Ion data. 
```

These e-expressions represent the invocation of user defined *macros* with arguments that are either themselves E-expressions
or value literals corresponding to the formal parameters of the macro's definition.
The resulting stream is then expanded into the resulting Ion data model.

### Macro definitions

Macros can be defined by a user either directly in a default module within an encoding directive or in a module
defined externally (i.e., shared module). A macro has a name which must be unique in a module *or* it may have no name.

In Ion binary, macros are always addressed in E-expressions by integer macro address. 
In Ion text, macros may be addressed by the offset in the local macro table (mirroring binary), by name, or by qualifying the macro name/offset with the module name in the encoding context.

> An E-expression can _only_ refer to macros installed in the local macro tabled.

Reviewer: ```suggestion
An E-expression can _only_ refer to macros installed in the local macro table.
```

**E-expressions name resolution**
```ion
// resolves to macro `bar` in the "default" module 
(:bar)

// resolves to macro `bar` in the "foo" module
(:foo::bar)

// resolves to macro 5 in the local macro table
(:5)
```

### Template definitions

User defined macros are defined by their *template* which defines how they are invoked and what data they evaluate to.

This template is defined as Ion data with a special-purpose E-Expression to signify a placeholder for an argument to be substituted.
Placeholders may accept any type of value, with an optional default value to use if no value is provided for that argument.
Other placeholders required an argument of a specific type—called ["tagless" values](macros/tagless_encodings.md)—whose 

> encodings do not begin with an opcode and are therefore both more compact and less flexible (For example: `int32`, `float16`).

Reviewer: ```suggestion
Placeholders for ["tagless" values](macros/tagless_encodings.md)—whose encodings do not begin with an opcode and are therefore more compact and less flexible than tagged values—require a type marker argument (e.g., `{#int32}`, `{#float16}`) to specify how the argument is encoded. 
```

### Directives

> A _directives_ is a top-level e-expression that modifies the encoding context.

Reviewer: ```suggestion
A _directive_ is a top-level e-expression that modifies the encoding context.
```

> In text, its uses the e-expression name `$ion`, and the first child value is an operation name.

Reviewer: ```suggestion
In text, directives use the e-expression name `$ion`, and the first child value is an operation name.
```

_Directives_ modify the encoding context.

> Syntactically, a directive is a top-level s-expression annotated with `$ion`.

Reviewer: This needs updating too

Its first child value is an operation name.
The operation determines what changes will be made to the encoding context and which values or clauses may legally follow.

```ion
(:$ion operation_name /*...*/ )
```

In Ion v1.1, there are eight supported directive operations:
1. [`module`](modules/directives.md#module-directives)
2. [`import`](modules/directives.md#import-directives)
3. [`encoding`](modules/directives.md#encoding-directives)
3. [`set_symbols`](modules/directives.md#set_symbols-directives)
3. [`add_symbols`](modules/directives.md#add_symbols-directives)
3. [`set_macros`](modules/directives.md#set_macros-directives)
3. [`add_macros`](modules/directives.md#add_macros-directives)

> 3. [`use`](modules/directives.md#use-directives)

Reviewer: Let's put these in opcode order and fix the numbering

1.0 shared symbol tables and Ion 1.1 shared modules in its catalog.

## Text syntax changes

Ion 1.1 text *must* use the `$ion_1_1` version marker at the top-level of the data stream or document.

Ion 1.1 introduces new syntax elements to represent e-expressions and tagless values.

The introduction of *encoding expression* (*E-expression*) syntax allows for the invocation of macros in the data stream. 
This syntax is grammatically similar to S-expressions, except that these expressions are opened with `(:` and closed with `)`.
For example, `(:a 1 2)` would expand the macro named `a` with the arguments `1` and `2`.
This syntax is allowed anywhere an Ion value is allowed. 
See the _[Macros, templates, and encoding expressions](#macros-templates-and-encoding-expressions)_ section for details.

Tagless-values are primarily a concern of the binary encoding, but there is a text encoding for them so that data can be
transcoded between text and binary without loss.

> The tag of a tagless value is represented in text as `{#<type>}`, where `<type>` can be any valid [Tagless Scalar Type](macros/tagless_encodings.md)

Reviewer: We've referred to these as "type markers" in some places, but we should standardize on a term.
